src/cargo-fmt/main.rs

```diff
@@ -445,10 +505,59 @@ fn get_targets_with_hitlist(
     }
 }
 
-fn add_targets(target_paths: &[cargo_metadata::Target], targets: &mut BTreeSet<Target>) {
+fn add_targets(
+    manifest_path: &PathBuf,
+    target_paths: &[cargo_metadata::Target],
+    targets: &mut BTreeSet<Target>,
+    include_nested_test_files: bool,
+) {
+    let mut test_files_added = false;
     for target in target_paths {
-        targets.insert(Target::from_target(target));
+        // Packages often have more than one `test` target,
+        // so only add the nested files for the first one.
+        let test_files = if include_nested_test_files
+            && !test_files_added
+            && target.kind.iter().any(|t| t == "test")
+        {
+            match manifest_path.parent() {
+                None => None,
+                Some(package_dir) => {
+                    let target_dir = package_dir.join("tests");
+                    test_files_added = true;
+                    Some(get_nested_integration_test_files(&target_dir, &target_dir))
+                }
+            }
+        } else {
+            None
+        };
+        targets.insert(Target::from_target(target, test_files));
+    }
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(path: &Path, root_dir: &Path) -> Vec<PathBuf> {
```

Comment:
We can change the signature to `Option<Vec<PathBuf>>` and handle errors by transferring them into `None`.